### ISSUE TITLE
somax: equivalent-barotropic SSH-QG and internal-tide shallow-water models (+ open/relaxation BCs)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/somax#141]
## TL;DR

Somax has a strong QG/SW model suite (`BarotropicQG`, `BaroclinicQG`, `ReparameterizedQG`, linear/nonlinear/multilayer SW, Lorenz, PDE toys). An audit against the MASSH SSH-mapping codebase ([`leguillf/MASSH@VarDyn`](https://github.com/leguillf/MASSH/tree/VarDyn)) found **two ocean models Somax lacks**, both central to altimetric SSH interpolation:

1. **Equivalent-barotropic (1.5-layer) QG parameterized directly in SSH** — the BFN-QG workhorse, with a finite-deformation-radius stretching term, spatially-varying Rossby radius, β, and a bathymetric PV term.
2. **Internal-tide linear shallow-water** — the 4DVar-SW model, with tidal-frequency forcing, equivalent depth, open boundaries, and baroclinic-mode coupling.

Plus a smaller item: **open / relaxation-to-external-field boundary conditions**.

> Note: an earlier draft of this audit also listed "SQG (surface QG)" as a gap. On verbatim review of `MASSH@VarDyn`, **no SQG model exists in that branch**, so it has been dropped. (SQG may still be worth doing on its own merits, but it is *not* a MASSH-derived gap.)

Scope: **model + numerical-primitive level only.** Assimilation drivers are out of scope (those live in pipekit-cycle / downstream).

---

## Background & motivation

Mapping SSH from sparse altimetry is dynamics-constrained interpolation: the assimilation propagates information in space/time through an ocean model. The model of choice determines the regime you can represent. MASSH ships exactly the two that Somax is missing.

### Gap 1 — equivalent-barotropic SSH-QG

MASSH `Qgm.h2pv` defines potential vorticity **directly from SSH `h`** (verbatim, `models/model_qg1l/jqgm.py`):

```python
q = g/f0 * ((h[2:,1:-1] + h[:-2,1:-1] - 2*h[1:-1,1:-1]) / dy**2
          + (h[1:-1,2:] + h[1:-1,:-2] - 2*h[1:-1,1:-1]) / dx**2) \
  - g*f0/(c**2) * h[1:-1,1:-1]
```

i.e. `q = (g/f₀)∇²η − (f₀/c²)·(g·η) = (g/f₀)∇²η − η·g·f₀/c²`, a **finite-deformation-radius** QG with the stretching term `−η/R_d²` where `R_d = c/f₀`. The inversion is a Helmholtz solve (verbatim):

```python
self.helmoltz_dst = g/f0 * laplace_dst - g*f0 / c**2     # formulation='ssh'
```

with **spatially-varying `c`** (the first-baroclinic phase speed → Rossby radius), an optional `bathymetry_PV_term`, β, open BCs, and TLM/adjoint (`step_tgl`/`step_adj`).

Somax's `BarotropicQG` is **not** this. From `somax/_src/models/qg/barotropic.py:129-137`, the state `q` is relative vorticity and inversion is **pure Poisson**:

```python
def _invert_pv(self, q):
    # Solves nabla^2 psi = q   (Dirichlet, psi=0 at walls)
    return streamfunction_from_vorticity(q, self.grid.dx, self.grid.dy, bc=self.poisson_bc)
```

There is no `1/R_d²` stretching, no SSH/η formulation, no topographic PV, and only constant `f0`/`beta`. So the standard altimetry-mapping QG is unrepresented. (`BaroclinicQG` can approximate it with `nl=2` modal Helmholtz, but there's no clean 1-layer SSH-QG with variable `R_d`.)

### Gap 2 — internal-tide linear shallow-water

MASSH `class CSWm` (verbatim, `models/model_sw1l/jswm.py`):

```python
def __init__(self, X=None, Y=None, dt=None, time_scheme='rk4', bc_kind='1d',
             g=9.81, f=1e-4, Heb=0.7, obc_south=True, obc_north=True,
             obc_west=True, obc_east=True, periodic_x=False, periodic_y=False,
             omegas=None, bc_theta=None, **arr_kwargs):
...
def compute_IT_2D(self, t, He, h_SN, h_WE, flag_tangent=True): ...
def _wave_phases(self, w, He, He_on_u, He_on_v): ...
def obcs(self, u, v, h, u0, v0, h0, He, w1ext): ...
def step_rk4(self, ...); def step_rk4_tgl(self, ...); def step_rk4_adj(self, ...)
```

with tidal frequencies `self.omegas` (e.g. M2, S2), equivalent depth `He = self.Heb if He is None else He + self.Heb`, boundary angle array `bc_theta`, open-boundary plane-wave internal-tide forcing (`compute_IT_2D` / `_wave_phases`), an Asselin-filtered leapfrog option, and TLM/adjoint.

Somax has linear/nonlinear/multilayer SW (`somax/_src/models/swm/{linear_1d,linear_2d,nonlinear_1d,nonlinear_2d,multilayer,linear_swm_terms}.py`), but **no tidal forcing, no equivalent-depth `He`, no open-boundary plane-wave forcing, no baroclinic-mode coupling** — so the internal-tide regime that 4DVar-SW targets is unrepresented.

### Already covered (not a gap, for the record)

MASSH `class QG(SW)` with `Q` / `G` / `QoG_inv` / `project_qg` (`P = G∘(Q∘G)⁻¹∘Q`, solver `dst_cmm`, `models/model_qgsw/qg.py`) is the **SW + idempotent geostrophic projection** formulation — which is exactly Somax's `ReparameterizedQG` (`somax/_src/models/qg/reparameterized.py`). Somax is level here.

---

## User stories

- **As an SSH-mapping researcher**, I want a 1-layer equivalent-barotropic QG keyed on SSH with a spatially-varying Rossby radius and bathymetric PV, so I can reproduce BFN-QG-style mapping inside the Somax/JAX stack and differentiate through it.
- **As an internal-tide modeler**, I want a linear SW model with tidal-frequency open-boundary forcing and equivalent depth so I can estimate coherent internal-tide SSH via 4DVar-SW.
- **As a DA practitioner**, I want both models to expose the standard Somax `SomaxModel` contract (`vector_field`, `diagnose`, `integrate`) so they drop into existing time-steppers and, via JAX autodiff, yield tangent-linear/adjoint for free.
- **As a coastal modeler**, I want open / relaxation-to-external-field boundary conditions so I can nest a Somax model inside a larger-domain boundary forcing.

---

## API proposal

Both models follow the existing `SomaxModel` pattern (eqx.Module state/params/consts/diagnostics, `vector_field`, `apply_boundary_conditions`, `diagnose`, `create` factory), reusing `finitevolx`/`spectraldiffx` primitives.

### Model 1 — `EquivalentBarotropicQG` (`somax/_src/models/qg/equivalent_barotropic.py`)

```python
class EquivalentBarotropicQGState(State):
    eta: Float[Array, "Ny Nx"]   # SSH (the prognostic field), or q — see note

class EquivalentBarotropicQGParams(Params):
    lateral_viscosity: Array        # nu
    bottom_drag: Array              # kappa

class EquivalentBarotropicQGConsts(PhysConsts):
    f0: float; beta: float; g: float = 9.81

class EquivalentBarotropicQG(SomaxModel):
    r"""1.5-layer equivalent-barotropic QG keyed on SSH.

        q = (g/f0) nabla^2 eta - (g*f0/c^2) eta + f0*h_b/H        (PV from SSH)
        dq/dt = -J(psi, q + beta*y) + nu*lap(q) - kappa*lap(psi)
        invert (nabla^2 - 1/Rd^2) psi-equivalent via Helmholtz (DST).

    Args:
        rossby_radius: c(x,y) first-baroclinic phase speed -> Rd = c/f0
            (scalar or 2-D field). Drives the finite-Rd stretching term.
        bathymetry_pv: optional topographic PV term f0*h_b/H.
    """
    rossby_phase_speed: Float[Array, "Ny Nx"] | float   # c
    bathymetry_pv: Float[Array, "Ny Nx"] | None = None
    helmholtz: HelmholtzCache                            # (lap - 1/Rd^2) solver

    def ssh_to_pv(self, eta): ...        # mirrors Qgm.h2pv
    def pv_to_ssh(self, q): ...          # Helmholtz inversion, mirrors Qgm.pv2h
    def ssh_to_uv(self, eta): ...        # geostrophy, mirrors Qgm.h2uv
    def vector_field(self, t, state, args=None): ...
```

Reuses `spectraldiffx.SpectralHelmholtzSolver2D` (the `(∇² − 1/R_d²)` solve maps directly onto MASSH's `helmoltz_dst = g/f0·laplace_dst − g·f0/c²`), `finitevolx.arakawa_jacobian`, and a 3rd-order upwind advection.

### Model 2 — `InternalTideSW` (`somax/_src/models/swm/internal_tide.py`)

```python
class InternalTideSWState(State):
    u: Float[Array, "Ny Nx"]; v: Float[Array, "Ny Nx"]; h: Float[Array, "Ny Nx"]

class InternalTideSWConsts(PhysConsts):
    f0: float; g: float = 9.81
    He: float = 0.7                       # equivalent depth (Heb)

class TidalForcing(eqx.Module):
    """Open-boundary plane-wave internal-tide forcing.
    Mirrors CSWm.compute_IT_2D / _wave_phases."""
    omegas: Float[Array, "n_tide"]        # tidal frequencies (M2, S2, ...)
    bc_theta: Float[Array, "n_bc"]        # boundary incidence angles
    def __call__(self, t, He): ...        # -> boundary u,v,h plane waves

class InternalTideSW(SomaxModel):
    r"""Linear rotating shallow water for coherent internal tides.

        du/dt =  f v - g dη/dx   (+ open-boundary tidal forcing)
        dv/dt = -f u - g dη/dy
        dη/dt = -He (du/dx + dv/dy)
    """
    forcing: TidalForcing
    open_bcs: tuple[bool, bool, bool, bool]   # S, N, W, E
    def vector_field(self, t, state, args=None): ...
    def apply_boundary_conditions(self, state): ...   # open / radiating
```

### BCs — open / relaxation to external field (`somax/_src/core/boundary.py` or extend domain)

```python
class RelaxationBC(eqx.Module):
    """Newtonian relaxation of boundary cells toward a (time-varying) external field.
    Mirrors MASSH src/bc.py Bc_ext + sponge."""
    external: Callable[[float], Array]    # boundary field provider
    weight: Float[Array, "Ny Nx"]         # sponge weight map Wbc
    def apply(self, state, t): ...
```

---

## API examples

**Equivalent-barotropic SSH-QG mapping a region with variable R_d:**

```python
import jax.numpy as jnp
from somax import EquivalentBarotropicQG

c = load_phase_speed(lon, lat)          # first-baroclinic c(x,y) -> Rd = c/f0
model = EquivalentBarotropicQG.create(
    nx=128, ny=128, Lx=1e6, Ly=1e6,
    f0=1e-4, beta=1.6e-11,
    rossby_phase_speed=c,               # spatially varying
    bathymetry_pv=topo_pv,              # optional
    lateral_viscosity=10.0,
)
state0 = model.ssh_to_state(eta_first_guess)
traj = model.integrate(state0, t0=0.0, t1=10*86400, dt=900.0)
u, v = model.ssh_to_uv(model.diagnose(traj[-1]).eta)   # geostrophic currents
```

**Internal-tide SW forced by M2 + S2 at open boundaries:**

```python
from somax import InternalTideSW, TidalForcing

M2 = 2*jnp.pi/(12.42*3600); S2 = 2*jnp.pi/(12.00*3600)
forcing = TidalForcing(omegas=jnp.array([M2, S2]), bc_theta=boundary_angles)
model = InternalTideSW.create(nx=256, ny=256, Lx=5e5, Ly=5e5,
                              f0=7e-5, He=0.7, forcing=forcing,
                              open_bcs=(True, True, True, True))
traj = model.integrate(model.zeros_state(), t0=0.0, t1=5*86400, dt=120.0)
ssh_it = traj[..., :].h    # coherent internal-tide SSH signature
```

**Autodiff gives TLM/adjoint for free** (so these drop into a 4D-Var driver):

```python
import jax
# tangent-linear = jvp, adjoint = vjp through model.integrate — no hand-coded step_tgl/step_adj
dssh = jax.jvp(lambda s: model.integrate(s, 0.0, T, dt), (s0,), (ds0,))
```

---

## TODO

- [ ] `qg/equivalent_barotropic.py`: `EquivalentBarotropicQG` (+ State/Params/Consts/Diagnostics, `create`).
  - [ ] `ssh_to_pv` with finite-`Rd` stretching term + optional bathymetric PV + β.
  - [ ] `pv_to_ssh` via `spectraldiffx` Helmholtz `(∇² − 1/Rd²)` (support spatially-varying `c`).
  - [ ] `ssh_to_uv` geostrophy; Arakawa Jacobian advection; 3rd-order upwind option.
- [ ] `swm/internal_tide.py`: `InternalTideSW` + `TidalForcing` (`omegas`, `bc_theta`, equivalent depth `He`, plane-wave OBC forcing).
- [ ] `core/boundary.py`: `RelaxationBC` / open-radiating BC (sponge weight map).
- [ ] Register both in `somax/_src/models/__init__.py` and public exports; add GFD test cases (`gfd_testcases.py`).
- [ ] Tests: PV↔SSH round-trip; geostrophic-balance check; energy/enstrophy invariants for the QG; gravity/inertial-wave dispersion + tidal-frequency response for the SW; finite-difference gradient check through `integrate` (autodiff TLM/adjoint).
- [ ] Docs notebooks: (a) equivalent-barotropic QG free run + geostrophic currents; (b) internal-tide SW forced by M2.
- [ ] Cross-check numerics against MASSH on a shared idealized case.

## Open questions

- For `EquivalentBarotropicQG`, prognose **SSH `η`** (as MASSH does) or **PV `q`** (as Somax's `BarotropicQG` does)? SSH-prognostic matches the altimetry use case and the MASSH formulation; PV-prognostic matches existing Somax conventions. Proposal: SSH-prognostic with `ssh_to_pv`/`pv_to_ssh` helpers.
- Should variable-`R_d` Helmholtz use `spectraldiffx` (constant-coefficient DST, with `c` frozen per-solve as MASSH does) or `finitevolx` variable-coefficient multigrid (`∇·(c∇)`)? MASSH uses the former.
- Multiple tidal constituents: model each frequency as a separate plane-wave field (MASSH) or a single forced field?

## References

- MASSH: https://github.com/leguillf/MASSH/tree/VarDyn
  - `models/model_qg1l/jqgm.py` — `Qgm` (`h2pv`, `pv2h`, `h2uv`, DST `helmoltz_dst`, `step_tgl`/`step_adj`)
  - `models/model_sw1l/jswm.py` — `CSWm` (`omegas`, `He`, `compute_IT_2D`, `_wave_phases`, `obcs`)
  - `models/model_qgsw/qg.py` — `QG(SW)` projection (≙ Somax `ReparameterizedQG`)
  - `src/bc.py` — `Bc_ext` external-field boundary nudging
- Current Somax models: `somax/_src/models/qg/barotropic.py` (`_invert_pv`, lines 129-137), `somax/_src/models/swm/`, `somax/_src/models/qg/reparameterized.py`
- Background:
  - Le Guillou et al. (2021), *Mapping altimetry in the forthcoming SWOT era by back-and-forth nudging a one-layer QG model.* JTECH.
  - Ubelmann, Klein & Fu (2015), *Dynamic interpolation of sea surface height and potential applications for future high-resolution altimetry mapping.* JTECH.
  - Le Guillou et al. (2021/2023), 4DVar estimation of internal-tide SSH from altimetry. Ocean Science.
  - Vallis (2017), *Atmospheric and Oceanic Fluid Dynamics*, ch. 5 (QG), ch. 3 (shallow water).
